### PR TITLE
Add logger specific to API (#2086)

### LINKED ISSUE #2086: Add custom logger for GRPC API
## Description
Right now, all of the new API code is using the default logger. We should either have a single logger shared by all the GRPC services, or else one instance per service.

A related change we may want to make here is reducing the default log level for many of the log statements, especially the ones in the reporter, from INFO to DEBUG (especially the "not reporting... since no one is listening" messages).

## Affected code
api/grpcserver/*

This issue appears in commit hash: latest in api-phase9 branch

### PR DESCRIPTION
**Pull request by hoenirvili**
## Motivation
All the **grpc** common functionality are placed in a single package and their all using the global logger. This is not good because what if the client want's to use a custom logger or other miscellaneous options?
 
## Changes
I've changed all files under api/grpcserver/* to make all the concrete implementation be flexible. By flexible I mean, using the option pattern and make all implementations support adding required miscellaneous things like support custom loggin, support grpc protocol options etc. Because I didn't wanted types like **optionNode**, **optionMesh**, etc. I've moved every concrete implementation into it's own package under grpcserver/* to make the naming once again sane and be more maintainable.

## Test Plan

I've just re-run the unit tests and made sure that everything passes alongside with whole compilation of the main binary (on linux 64bit)

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

### CODE REVIEW

**Comment by y0sher** [comment]
Hi @hoenirvili , the whole thing looks great. I would really like @lrettig to have a look since he is working on the API atm. 
also one thing I can tell is that this PR is missing some functionality we have for loggers - controlling the log level dynamically. 
if you look on the top of the `node.go` flie you'd find constants we use to name each log, this is created by maintaining a map with the state and loading it from config or changed from the grpc api. please have a look and try to add it, contact me if you need any help.

File: cmd/node/node.go
Comment (noamnelke) [inline cmd/node/node.go:61]: the comment should be on the same line as the import - it explains it

**Review by noamnelke — COMMENTED** [review]
some of the file renames were correctly understood by git and others weren't - can you please fix this so it's easier to see the changes in the moved files, and to maintain the correct history of the files?

**Comment by noamnelke** [comment]
Thinking about this a bit more, I'm not sure the relatively big refactor and partition into packages should be in the same PR as the logger change. @lrettig ?

File: cmd/node/node.go
Comment (noamnelke) [reply cmd/node/node.go:61]: note that this also causes linting to fail

**Comment by noamnelke** [comment]
Note that `TestSpacemeshApp_GrpcService` panics.

File: cmd/node/node.go
Comment (hoenirvili) [reply cmd/node/node.go:61]: Thanks, I did move the comment back where it was. I think go-fmt moved the single **import** into the big import and didn't moved the comment.

**Comment by hoenirvili** [comment]
> 
> 
> Hi @hoenirvili , the whole thing looks great. I would really like @lrettig to have a look since he is working on the API atm.
> also one thing I can tell is that this PR is missing some functionality we have for loggers - controlling the log level dynamically.
> if you look on the top of the `node.go` flie you'd find constants we use to name each log, this is created by maintaining a map with the state and loading it from config or changed from the grpc api. please have a look and try to add it, contact me if you need any help.

Ok, I think I understood what you mean. I did checked the code and saw the map of loggers and I saw that this is configurable and you can set dynamically the log level across multiple resources so I think I managed to add support/port the functionality to all services/servers types.

Can you check one more time and let me know if this is what you want?

File: cmd/node/node.go
Comment (y0sher) [inline cmd/node/node.go]: didn't you mean to use `addLogger` here as well?

File: cmd/node/node.go
Comment (y0sher) [inline cmd/node/node.go]: the method that calls `startAPIServices` (= `Start`) is creating a logger that way already. it will be better to pass it down to this instead of creating a new one. though if you see a lot of clashes in apis and other usages than we can live with this duplication. though I'm pretty sure we only use it once so should be easy to pass down the root logger..

**Comment by y0sher** [comment]
> > Hi @hoenirvili , the whole thing looks great. I would really like @lrettig to have a look since he is working on the API atm.
> > also one thing I can tell is that this PR is missing some functionality we have for loggers - controlling the log level dynamically.
> > if you look on the top of the `node.go` flie you'd find constants we use to name each log, this is created by maintaining a map with the state and loading it from config or changed from the grpc api. please have a look and try to add it, contact me if you need any help.
> 
> Ok, I think I understood what you mean. I did checked the code and saw the map of loggers and I saw that this is configurable and you can set dynamically the log level across multiple resources so I think I managed to add support/port the functionality to all services/servers types.
> 
> Can you check one more time and let me know if this is what you want?

Exactly what I wanted, added a few minor comments. still waiting for @lrettig to take a look here before merging.

File: cmd/node/node.go
Comment (hoenirvili) [reply cmd/node/node.go]: I've noticed that meshlogger it's already set in the map so I figured I can just get ptr from the map and just pass/construct only the outer required thing. I can change to addLogger but it will overwrite the same mesh logger or we could add a new var key with a different name. You're call.

File: cmd/node/node.go
Comment (hoenirvili) [reply cmd/node/node.go]: Ok, make sense. I will reuse the app.Logger I thought by using the same initial constructed one would not work. Thanks for the valuable insight.

**Comment by hoenirvili** [comment]
> Exactly what I wanted, added a few minor comments. still waiting for @lrettig to take a look here before merging.

I addressed the comments, how is it now?

**Comment by y0sher** [comment]
> > Exactly what I wanted, added a few minor comments. still waiting for @lrettig to take a look here before merging.
> 
> I addressed the comments, how is it now?

The logger attached to `app` is already named and is used for app logs. I'd rather we pass down the root logger created in the function that calls `startAPIServices`.
also sorry I missed it earlier but for the log level config to work you will also have to add config entries for your loggers and add them in `addLogger`.

**Comment by hoenirvili** [comment]
> 
> 
> > > Exactly what I wanted, added a few minor comments. still waiting for @lrettig to take a look here before merging.
> > 
> > 
> > I addressed the comments, how is it now?
> 
> The logger attached to `app` is already named and is used for app logs. I'd rather we pass down the root logger created in the function that calls `startAPIServices`.
> also sorry I missed it earlier but for the log level config to work you will also have to add config entries for your loggers and add them in `addLogger`.

Thanks once again for the patience, I just modified the code. How is it now?

Btw, I have a proposal, instead of relaying on that switch stmt why won't we just have a map of these keys like ```map[keys]bool``` or ```map[keys]struct{}``` and if an element is found then it will return ok and we can go and use the **logger**. This way we only have to modify in one place the **map** and everything will work and also we also have the benefit at runtime to add dynamic support for *future keys*. 

Do you want me in this **PR** to add this or make a new one after this is merged and continue further improving this ?

File: api/pb/api.proto
Comment (lrettig) [inline api/pb/api.proto]: This file, and all the included protobuf files, will be deprecated very soon - let's not bother updating them, okay?

File: cmd/node/node.go
Comment (lrettig) [inline cmd/node/node.go]: Let's pick a consistent naming convention for the loggers. I like `meshService` for the mesh service, how about `nodeService` for this one?

File: cmd/node/node.go
Comment (lrettig) [inline cmd/node/node.go:170]: It's not as immediately obvious from the package name `server` what this package does, `grpcserver` is a bit clearer and more intuitive isn't it?

**Review by lrettig — CHANGES_REQUESTED** [review]
Thanks for the PR! Added a few quick comments. GitHub's diff looks off for the renamed files, I will take a closer look at those later. Could you rebase against the `api-phase9` branch? We'll be merging that branch soon, and this will prevent some ugly merge conflicts when that happens. #2092 is also in flight and will land soon and might cause some conflicts here, so we might want to merge that in here as well to prevent downstream conflicts.

**Comment by lrettig** [comment]
> Thinking about this a bit more, I'm not sure the relatively big refactor and partition into packages should be in the same PR as the logger change. @lrettig ?

Agree here, it would be nice to split up these concerns - the partitioning into packages isn't required for the logging is it?

**Comment by lrettig** [comment]
> some of the file renames were correctly understood by git and others weren't - can you please fix this so it's easier to see the changes in the moved files, and to maintain the correct history of the files?

I'm seeing this issue now as well, odd - @hoenirvili is this something you might be able to fix? I'm not sure why GH does this.

File: cmd/node/node.go
Comment (lrettig) [reply cmd/node/node.go]: These are definitely two different things - `MeshLogger` should be distinct from `MeshServiceLogger`, which is what you want here. The GRPC services should get distinct loggers with the name `<Svcname>ServiceLogger`.

File: cmd/node/node.go
Comment (lrettig) [reply cmd/node/node.go]: > Btw, I have a proposal, instead of relaying on that switch stmt why won't we just have a map of these keys like `map[keys]bool` or `map[keys]struct{}` and if an element is found then it will return ok and we can go and use the **logger**. This way we only have to modify in one place the **map** and everything will work and also we also have the benefit at runtime to add dynamic support for _future keys_.
> 
> Do you want me in this **PR** to add this or make a new one after this is merged and continue further improving this ?

I don't have an opinion on replacing the switch with a map, will defer to @y0sher and @noamnelke on that. But if we do make that change, let's do it in a separate branch/PR - small PRs are good! :)

File: cmd/node/node.go
Comment (hoenirvili) [reply cmd/node/node.go:170]: Ok, I will change this into `grpcserver.Server`

**Comment by y0sher** [comment]
> > > > Exactly what I wanted, added a few minor comments. still waiting for @lrettig to take a look here before merging.
> > > 
> > > 
> > > I addressed the comments, how is it now?
> > 
> > 
> > The logger attached to `app` is already named and is used for app logs. I'd rather we pass down the root logger created in the function that calls `startAPIServices`.
> > also sorry I missed it earlier but for the log level config to work you will also have to add config entries for your loggers and add them in `addLogger`.
> 
> Thanks once again for the patience, I just modified the code. How is it now?
> 
> Btw, I have a proposal, instead of relaying on that switch stmt why won't we just have a map of these keys like `map[keys]bool` or `map[keys]struct{}` and if an element is found then it will return ok and we can go and use the **logger**. This way we only have to modify in one place the **map** and everything will work and also we also have the benefit at runtime to add dynamic support for _future keys_.
> 
> Do you want me in this **PR** to add this or make a new one after this is merged and continue further improving this ?

Good job on fixing it. the only part missing are the entries in the `config.toml` file for future reference. 
It would be great if you could try and improve this feature but I believe it is better to do so in a separate PR.

**Comment by hoenirvili** [comment]
> 
> 
> > > > > Exactly what I wanted, added a few minor comments. still waiting for @lrettig to take a look here before merging.
> > > > 
> > > > 
> > > > I addressed the comments, how is it now?
> > > 
> > > 
> > > The logger attached to `app` is already named and is used for app logs. I'd rather we pass down the root logger created in the function that calls `startAPIServices`.
> > > also sorry I missed it earlier but for the log level config to work you will also have to add config entries for your loggers and add them in `addLogger`.
> > 
> > 
> > Thanks once again for the patience, I just modified the code. How is it now?
> > Btw, I have a proposal, instead of relaying on that switch stmt why won't we just have a map of these keys like `map[keys]bool` or `map[keys]struct{}` and if an element is found then it will return ok and we can go and use the **logger**. This way we only have to modify in one place the **map** and everything will work and also we also have the benefit at runtime to add dynamic support for _future keys_.
> > Do you want me in this **PR** to add this or make a new one after this is merged and continue further improving this ?
> 
> Good job on fixing it. the only part missing are the entries in the `config.toml` file for future reference.
> It would be great if you could try and improve this feature but I believe it is better to do so in a separate PR.


I just did that now, I still need to figure out how to sync with `api-phase9` and it should be fine after that I think.

File: cmd/node/node.go
Comment (hoenirvili) [reply cmd/node/node.go]: Deal.

File: api/pb/api.proto
Comment (hoenirvili) [reply api/pb/api.proto]: Ok, I didn't know that. The reason I added these line to get rid of the annoying warning from generating the proto files.

File: cmd/node/node.go
Comment (hoenirvili) [reply cmd/node/node.go]: Ok, sounds good to me !

File: cmd/node/node.go
Comment (hoenirvili) [reply cmd/node/node.go]: I did used/updated all naming logger acordignly.

File: api/pb/api.proto
Comment (lrettig) [reply api/pb/api.proto]: No worries! The files are going away because we have a full golang build of the new API in https://github.com/spacemeshos/api/tree/master/release/go, btw, and we can use it directly as a library rather than compiling the protobuf inside go-spacemesh, which is much cleaner (and better separation of concerns).

**Comment by lrettig** [comment]
> I just did that now, I still need to figure out how to sync with api-phase9 and it should be fine after that I think.

Step one: create a backup of your current branch in case the rebase goes wrong :)

`> git checkout -b mybranch-backup && git checkout -`

Step two: rebase against api-phase9. Hopefully this is relatively painless.

`> git rebase origin/api-phase9`

Step three: force push your changes

`> git push -f`

Step four: [change the base branch](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request) of this PR.

Some more possibly helpful info here:

- https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request
- https://anavarre.net/how-to-rebase-a-github-pull-request/
- https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

**Comment by hoenirvili** [comment]
> 
> 
> > I just did that now, I still need to figure out how to sync with api-phase9 and it should be fine after that I think.
> 
> Step one: create a backup of your current branch in case the rebase goes wrong :)
> 
> `> git checkout -b mybranch-backup && git checkout -`
> 
> Step two: rebase against api-phase9. Hopefully this is relatively painless.
> 
> `> git rebase origin/api-phase9`
> 
> Step three: force push your changes
> 
> `> git push -f`
> 
> Step four: [change the base branch](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request) of this PR.
> 
> Some more possibly helpful info here:
> 
>     * https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request
> 
>     * https://anavarre.net/how-to-rebase-a-github-pull-request/
> 
>     * https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

Thanks man you're awesome, tho' I knew what I had to do (as in git commands), just I'm lazy to resolve all the conflicts that appeared after rebase. I will fix this asap tho.'

**Comment by lrettig** [comment]
Sounds good. Hopefully there aren't too many conflicts! Sorry to create
extra work for you

**Comment by lrettig** [comment]
Thanks for rebasing, I'm sure that was a lot of work. I've begun reviewing your PR but it's a very large refactor and it may take me a few days to review it completely.

**Comment by hoenirvili** [comment]
What about now? Is the patch fine?

**Comment by hoenirvili** [comment]
So what's the right decision, I noticed that `api-phase9` is merged with this branch. Should I point now the PR to `api-phase9`and when `api-phase9` is ready it will be merged to `master` ?

Or do you want me this split this into multiple PR's and start sending one at a time to `develop` ?

**Comment by lrettig** [comment]
It looks like this was closed automatically since the base branch was merged into develop. I can't reopen this. We'll need to transfer this over to a new PR against develop.

**Comment by lrettig** [comment]
> Or do you want me this split this into multiple PR's and start sending one at a time to develop ?

Yes let's do this! :)

**Pull request closed without merging**
